ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/castor/algorithm/CastorFullOptimization.java

```diff
@@ -736,13 +730,138 @@ static void applyPreventiveResultsForAutoOrCurativeRangeActions(Network network,
+            contingencyResults.forEach((state, result) -> {
+                if (isRangeActionAvailableInState(ra, state, crac)) {
+                    setPointResults.putIfAbsent(state, new HashMap<>());
+                    setPointResults.get(state).put(ra, result.getOptimizedSetpoint(ra, state));
+                }
+            });
+        });
+        return setPointResults;
+    }
+
+    /**
+     * Checks if raUsageLimits are at risk if we choose to re-optimize a range action.
+     * Returns True if it's at risk, False otherwise.
+     */
+    private static boolean shouldRemoveRaDueToUsageLimits(String operator, RaUsageLimits raUsageLimits, Set<RangeAction<?>> activatableRangeActions, Set<NetworkAction> activatedNetworkActions) {
+        if (operator == null) {
+            return raUsageLimits.getMaxRa() < activatableRangeActions.size() + activatedNetworkActions.size();
+        }
+
+        Set<RemedialAction<?>> activatableRemedialActions = new HashSet<>(activatableRangeActions);
+        activatableRemedialActions.addAll(activatedNetworkActions);
+
+        long activatableRangeActionsForTheTso = activatableRangeActions.stream().filter(ra -> operator.equals(ra.getOperator())).count();
+        long activatableRemedialActionsForTheTso = activatableRemedialActions.stream().filter(ra -> operator.equals(ra.getOperator())).count();
+        long activatableTsos = activatableRemedialActions.stream().map(RemedialAction::getOperator).filter(Objects::nonNull).distinct().count();
+
+        int limitingRangeActionValueForTheTso = raUsageLimits.getMaxPstPerTso().getOrDefault(operator, Integer.MAX_VALUE);
+        int limitingRemedialActionValueForTheTso = raUsageLimits.getMaxRaPerTso().getOrDefault(operator, Integer.MAX_VALUE);
+
+        return raUsageLimits.getMaxRa() < activatableRangeActions.size() + activatedNetworkActions.size()
+            || limitingRangeActionValueForTheTso < activatableRangeActionsForTheTso
+            || limitingRemedialActionValueForTheTso < activatableRemedialActionsForTheTso
+            || raUsageLimits.getMaxTso() < activatableTsos;
+    }
+
+    /**
+     * Gathers every range action that should not be considered in the second preventive if those 2 criterion are met :
+     * 1- The range action has the same tap in preventive and in a contingency scenario.
+     * 2- For the given state, the crac has limiting RaUsageLimits.
+     */
+    private static Set<RangeAction<?>> getRangeActionsToRemove(Crac crac, State preventiveState, Map<State, Map<RangeAction<?>, Double>> setPointResults, Map<State, OptimizationResult> contingencyResults) {
+        Set<RangeAction<?>> rangeActionsToRemove = new HashSet<>();
+        setPointResults.forEach((state, spMap) -> {
+            if (!state.isPreventive()) {
+                Set<RangeAction<?>> activatableRangeActions = crac.getPotentiallyAvailableRangeActions(state);
```

Comment:
Maybe rename this to availableRangeActions

Reply:
I've used `activatable` instead of  `available` because the real term is `potentiallyAvailable` and I wanted to keep in mind the potentiality. But I think elsewhere in the code we use `available` so why not I'm not